.semgrep/fsm_time.yml

```
@@ -0,0 +1,26 @@
+rules:
+  - id: "no-time-in-fsm"
+    patterns:
+      - pattern: time.Now()
+
+      # Metric state is local to the server and therefore must use time.
+      - pattern-not-inside: |
+          defer metrics.MeasureSince(..., time.Now())
+
+      # The timetable's whole point is to roughly track timestamps for Raft log
+      # indexes, so it must use time.
+      - pattern-not-inside: |
+          $N.timetable.Witness(log.Index, time.Now().UTC())
```

```suggestion
      # Metric state is local to the server and therefore must use time.
      - pattern-not-inside: |
          defer metrics.MeasureSince(...)

      # The timetable's whole point is to roughly track timestamps for Raft log
      # indexes, so it must use time.
      - pattern-not-inside: |
          $N.timetable.Witness(...)
```

I think it doesn't really matter what parameters are passed to these functions?

---

Perfect